File: algorithms/2_medium/O/00553_optimal_division/solution.c
static void push_num_into_vec(char* vec, int* vec_len, int num)
{
    assert(num >= 2 && num <= 1000);
    const int start_idx = *vec_len;
    while (num > 0) {
        vec[*vec_len] = (num % 10) + '0';
        *vec_len += 1;
        num /= 10;
    }
    int lo = start_idx;
    int hi = *vec_len - 1;
    while (lo < hi) {
        const int temp = vec[lo];
        vec[lo] = vec[hi];
        vec[hi] = temp;

        lo += 1;
        hi -= 1;
    }
}

char* optimalDivision(int* nums, const int len)
{
    assert(len > 0 && len <= 10 && nums != NULL);
    char* res = malloc(60 * sizeof(*res));
    assert(res != NULL);
    int res_len = 0;

    switch (len) {
        case 1:
            push_num_into_vec(res, &res_len, nums[0]);
            break;
        case 2:
            push_num_into_vec(res, &res_len, nums[0]);
            res[res_len] = '/';
            res_len += 1;
            push_num_into_vec(res, &res_len, nums[1]);
            break;
        default:
            push_num_into_vec(res, &res_len, nums[0]);
            res[res_len] = '/';
            res_len += 1;
            res[res_len] = '(';
            res_len += 1;
            push_num_into_vec(res, &res_len, nums[1]);
            for (int i = 2; i < len; i +=1) {
                res[res_len] = '/';
                res_len += 1;
                push_num_into_vec(res, &res_len, nums[i]);
            }
            res[res_len] = ')';
            res_len += 1;
    }

    res[res_len] = '\0';
    return res;
}

File: algorithms/2_medium/O/00553_optimal_division/solution.rs
impl Solution {
    fn push_num_into_vec(vec: &mut Vec<u8>, mut num: i32) {
        assert!(num >= 2);
        let start_idx = vec.len();
        while num > 0 {
            vec.push((num % 10) as u8 + b'0');
            num /= 10;
        }
        vec[start_idx..].reverse();
    }

    pub fn optimal_division(nums: Vec<i32>) -> String {
        let mut res = Vec::with_capacity(60);

        match nums.len() {
            0 => unreachable!(),
            1 => {
                Self::push_num_into_vec(&mut res, nums[0]);
            },
            2 => {
                Self::push_num_into_vec(&mut res, nums[0]);
                res.push(b'/');
                Self::push_num_into_vec(&mut res, nums[1]);
            },
            _ => {

                Self::push_num_into_vec(&mut res, nums[0]);
                res.push(b'/');
                res.push(b'(');
                Self::push_num_into_vec(&mut res, nums[1]);
                for num in nums.into_iter().skip(2) {
                    res.push(b'/');
                    Self::push_num_into_vec(&mut res, num);
                }
                res.push(b')');
            }
        }

        String::from_utf8(res).unwrap()
    }
}
